### PR TITLE
Introduce scheduled executor wrapper with dynamic interval

### PR DESCRIPTION
**Pull request by vishesh92**
### Description

This PR adds the ability for a Scheduled execution timing to be adjusted dynamically. Normally, adjusting the interval that these run requires a restart of the management services, which can be problematic for experimenting with intervals or just making adjustments without downtime.

Introduces `ConfigKeyScheduledExecutionWrapper` that can read a `Long` from a `ConfigKey` and dynamically adjust execution interval. The current implementation only supports emulating the behavior of `scheduleAtFixedRate` - meaning if the first run is at `N`, the next is at `N + configKey.value()`, ignoring the run time of the task itself. If the task takes longer than `configKey.value()`, the next run occurs immediately.

This PR also converts `ResourceCountCheckTask` to run on dynamic interval.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [x] Minor
- [ ] Trivial

### How Has This Been Tested?

Unit tests and manual testing.

I tested the following sceenarios:
1. Check that the task gets executed at the specified intervals (- time of execution).
2. If the value is set to a value <= 0, if the task is running, it will complete it's current execution and after that it gets disabled.
3. On changing the value from disabled to enabled, it gets enabled after 60 seconds (`enableIntervalSeconds`).
4. On decreasing the value, the new configuration takes effect after the current value of the interval.

### CODE REVIEW

**Comment by vishesh92** [comment]
@blueorangutan package

**Review by borisstoyanov — APPROVED** [review]
LGTM, manually tested

**Comment by GutoVeronezi** [comment]
> LGTM, manually tested

@vishesh92 @borisstoyanov could you be more clear about the tests you have done?

**Comment by vishesh92** [comment]
@blueorangutan package

**Comment by vishesh92** [comment]
> > LGTM, manually tested
> 
> @vishesh92 @borisstoyanov could you be more clear about the tests you have done?

I have updated the description of the PR with the cases I tested.

File: framework/config/src/main/java/org/apache/cloudstack/framework/config/ConfigKeyScheduledExecutionWrapper.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+        return this.executorService.schedule(this, duration, this.unit);
+    }
+
+    long getConfigValue() {
+        if (this.configKey.value() instanceof Long) {
+            return (Long) this.configKey.value();
+        } else if (this.configKey.value() instanceof Integer) {
+            return (Integer) this.configKey.value();
+        } else {
+            throw new IllegalArgumentException("ConfigKey value must be a Long or Integer");
```
Comment (DaanHoogland) [inline framework/config/src/main/java/org/apache/cloudstack/framework/config/ConfigKeyScheduledExecutionWrapper.java:88]: why are both long and interger allowed? isn't it easier to just define the datatype?

File: framework/config/src/main/java/org/apache/cloudstack/framework/config/ConfigKeyScheduledExecutionWrapper.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        }
+    }
+
+    @Override
+    public void run() {
+        if (getConfigValue() <= 0) {
+            executorService.schedule(this, enableIntervalSeconds, TimeUnit.SECONDS);
```
Comment (DaanHoogland) [inline framework/config/src/main/java/org/apache/cloudstack/framework/config/ConfigKeyScheduledExecutionWrapper.java:95]: what happens if `ResourceCountCheckInterval` was set to a negative value before starting the management server? Should we defend against that?

**Review by DaanHoogland — COMMENTED** [review]
clgtm generally , one question and one worry though

**Comment by vishesh92** [comment]
@blueorangutan package

File: framework/config/src/main/java/org/apache/cloudstack/framework/config/ConfigKeyScheduledExecutionWrapper.java (hunk 2 above)
Comment (vishesh92) [reply framework/config/src/main/java/org/apache/cloudstack/framework/config/ConfigKeyScheduledExecutionWrapper.java:95]: If the value is negative, we use `enableIntervalSeconds` as the duration for schedule.

File: framework/config/src/main/java/org/apache/cloudstack/framework/config/ConfigKeyScheduledExecutionWrapper.java (hunk 1 above)
Comment (vishesh92) [reply framework/config/src/main/java/org/apache/cloudstack/framework/config/ConfigKeyScheduledExecutionWrapper.java:88]: This is to allow usage of this class for both integer & long type config.

**Review by sureshanaparti — APPROVED** [review]
code lgtm

**Comment by sureshanaparti** [comment]
@blueorangutan package

File: framework/config/src/main/java/org/apache/cloudstack/framework/config/ConfigKeyScheduledExecutionWrapper.java (hunk 2 above)
Comment (DaanHoogland) [reply framework/config/src/main/java/org/apache/cloudstack/framework/config/ConfigKeyScheduledExecutionWrapper.java:98]: Ok, I was confusing the two constructors, I see the protected one at line 69 is getting a `enableIntervalSeconds` but doesn't check its value. I think a guard against negative values there would help.

On line 228 from `ResourceLimitManagerImpl` the constructor with the default `enableIntervalSeconds` is used. Can we remove the other constructor? I don't see any use of it.

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by borisstoyanov** [comment]
> > LGTM, manually tested
> 
> @vishesh92 @borisstoyanov could you be more clear about the tests you have done?

I've set the GS to 10 seconds and verified the job executes accordingly, then did some boundary testing and verified the schedule is disabled with setting value of 0 and negative. 
```
13:00:22,874 INFO  [c.c.r.ResourceLimitManagerImpl] (ResourceCountChecker-1:ctx-2a6bc5af) (logid:7041d095) Started resource counters recalculation periodic task.
13:00:24,689 INFO  [c.c.r.ResourceLimitManagerImpl] (ResourceCountChecker-1:ctx-2a6bc5af) (logid:7041d095) Finished resource counters recalculation periodic task.
13:00:32,921 INFO  [c.c.r.ResourceLimitManagerImpl] (ResourceCountChecker-1:ctx-2239d7a9) (logid:74570f70) Started resource counters recalculation periodic task.
13:00:33,395 INFO  [c.c.r.ResourceLimitManagerImpl] (ResourceCountChecker-1:ctx-2239d7a9) (logid:74570f70) Finished resource counters recalculation periodic task.
13:00:42,926 INFO  [c.c.r.ResourceLimitManagerImpl] (ResourceCountChecker-1:ctx-92e12d7e) (logid:ad039d2b) Started resource counters recalculation periodic task.
13:00:43,431 INFO  [c.c.r.ResourceLimitManagerImpl] (ResourceCountChecker-1:ctx-92e12d7e) (logid:ad039d2b) Finished resource counters recalculation periodic task.
13:00:52,985 INFO  [c.c.r.ResourceLimitManagerImpl] (ResourceCountChecker-1:ctx-bc300f8a) (logid:01944963) Started resource counters recalculation periodic task.
13:00:53,732 INFO  [c.c.r.ResourceLimitManagerImpl] (ResourceCountChecker-1:ctx-bc300f8a) (logid:01944963) Finished resource counters recalculation periodic task.
13:01:02,958 INFO  [c.c.r.ResourceLimitManagerImpl] (ResourceCountChecker-1:ctx-9e5a7d0d) (logid:54c8cf4b) Started resource counters recalculation periodic task.
13:01:03,492 INFO  [c.c.r.ResourceLimitManagerImpl] (ResourceCountChecker-1:ctx-9e5a7d0d) (logid:54c8cf4b) Finished resource counters recalculation periodic task.
13:01:12,975 INFO  [c.c.r.ResourceLimitManagerImpl] (ResourceCountChecker-1:ctx-cc951ced) (logid:ef37a93f) Started resource counters recalculation periodic task.
13:01:13,509 INFO  [c.c.r.ResourceLimitManagerImpl] (ResourceCountChecker-1:ctx-cc951ced) (logid:ef37a93f) Finished resource counters recalculation periodic task.
```

**Pull request merged**